packages/core-app-api/src/apis/implementations/SignalsApi/SignalsClient.ts

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { io, Socket } from 'socket.io-client';
+import { ConfigApi, IdentityApi, SignalsApi } from '@backstage/core-plugin-api';
+
+type SignalsSubscription = {
+  pluginId: string;
+  topic?: string;
+  onMessage: (data: any) => void;
+};
+
+export type SignalsMessage = {
+  pluginId: string;
+  topic?: string;
+  data: unknown;
+};
+
+/**
+ * An default implementation of the signals API
+ *
+ * @public
+ */
+export class SignalsClient implements SignalsApi {
+  private endpoint: string;
+  private ws?: Socket;
+  private signalsEnabled: boolean;
+  private subscriptions: Map<string, SignalsSubscription>;
+  private readonly messageQueue: Set<{ channel: string; message: any }>;
+  private readonly identity: IdentityApi;
+
+  static create(options: { configApi: ConfigApi; identityApi: IdentityApi }) {
+    const { configApi, identityApi } = options;
+    return new SignalsClient(configApi, identityApi);
+  }
+
+  private constructor(config: ConfigApi, identity: IdentityApi) {
+    this.subscriptions = new Map();
+    this.messageQueue = new Set();
+    this.signalsEnabled = false;
+    this.endpoint = '';
+    this.identity = identity;
+    this.readConfig(config);
+  }
+
+  private readConfig(config: ConfigApi) {
+    const baseUrl =
+      config.getOptionalString('backend.baseUrl') ?? 'http://localhost/';
```

Comment:
Talked a bit about this elsewhere, but this is not an assumption that we can make. We need to use the `DiscoveryApi` to grab the backend base URL, and for that we'll also need a plugin ID.